### ISSUE TITLE
items without listeners should not hitTest – onMouseEnter, onMouseLeave

### ISSUE THREAD

**Issue by dmccuskey** [paperjs/paper.js#1309]
i'm creating a charting application. i have display bars, etc on a background. the background is watching for events : mouseEnter/Leave/Move.

**Observed Behavior**

when the mouse moves over one of the chart items, that item captures/interrupts mouse events, even though those items have no attached listeners.

**Expected Behavior**

i would expect objects that DO NOT have attached event listeners to essentially be "invisible" to mouse activity. 

here's [a sketch of the issue](http://sketch.paperjs.org/#S/jZI9T8MwEIb/yslL06pKYWCpxFQhlrLQETM4yTWxetwF20lVVf3v2A4fXUAoH7Z17/u8uVPOis0bqrXaHTDUnVqqWpp0Ho0DMid0cA+MR9imfTHXrDmVWtd/FnrToysfnQx9KmdPaZpm01lqiizMrskn0TU5dl1cymesg+GWsMiwq+PNMt23d9ML5okipcdQnCMs7C3RRkjcGmYnJJLjTPMlamLgVbxMNuEnGTw+cMgN7QeugxUuAEfkEOGZWQt7ISxJ2kKrLNYqAS5XjC2aEf/LyOJvRrpWC9Bc0YDg3wfj4tLJQA2wBLApcegDYNqsKJvZW64RbJj5LCLrA7LlFvbipmwfq1+cCkEry6P1tqIYrXmxil//77Gnef88f449dfH70OOvVDk0h15iW16tX14vHw==)

**Otherwise...**

_Option 1_
 
from what i've read elsewhere, i guess all items are getting hit-tested. i see in `Item.js` that the following are in `_hitTest()`:

```
        if (this._locked || !this._visible || this._guide && !options.guides
                || this.isEmpty()) {
            return null;
        }
```

so it appears that i could set `._locked`, but i can't find an "official" way in the code to do so.

_Option 2_

i could attach listeners to all of my chart items and re-emit the events, but i could have hundreds of items on the graph. this option seems memory and performance inefficient.

_Option 3_

i'm open to suggestions...

**Comment by ubwhole** [comment]
@dmccuskey When I log the ID of the shape, it logs the ID of the blue rectangle.  But when I assign the blue rectangle shape to a new variable, it logs the ID of the yellow rectangle.

```
...
console.log("Created new yellow rectangle shape " + o.id)
grp.addChild( o )
o.onMouseEnter = function( event ) {
	console.log("Enter " + o.id )
}
...
```
Is assigning to a new variable the correct thing to do?

**Comment by dmccuskey** [comment]
@ubwhole i've found that inside of the event-handling function,`this` is set to the proper object. in my example i'm just using `o` as a temporary variable for setup.

```
o.onMouseLeave = function( event ) {
	console.log("Leave", this.type, this.fillColor )
}
```

here is an [updated sketch](http://sketch.paperjs.org/#S/jVM7T8MwEP4rJy9NqyqBgaUSE0IsZaEjZnCSa2LV3AXbSVVV+e/YSYEilRLl4bPue/jyKUdB6h3FSmx26ItaLEXBZdx3yoJRB7RwD4R7WMc6mUuSFFuVbU6NRjVo0yfLbRPbAydVZflQa1MmA3BgjTwOrJGxqcOSvmDhFVUGk0HsbHuzjPft3fiCeVTh1KFPjkHMb7UxD2zYrmB2QGN4P5PUB0wwPLPnkcb0zK3DR/LDQNuWCq+ZEsAOyQfxQbNgcmwwNVwlUozgUVuKJfhau9QfGjyV3yeIFv2ZyxpVh1NdRvB0l3hlC5CUmxbBfbTKhqXm1pRA7EHHU7eNB4xFZqK8JKepQNB+5gaQ0c4jaapgG4SH07nQ/dLJEaTQ1Gmnc4NSSFpkYb7J0cXMfp6r0cUp/g4uy2BadgF4Ob6o/+9nDez+t9nVCC+YjfjJZuE3yy2qXcMhLidWr2/9Jw==)

as you'll see, i have commented out the event handlers for the blue rectangle.

**Comment by ubwhole** [comment]
@dmccuskey Using the `this` object makes more sense.  So the problem lies with crossing the border between the blue and yellow box.  Correct?

**Comment by dmccuskey** [comment]
@ubwhole yes the issue is when you cross over into the blue box. if there isn't an event handler on that object, then that object should be "invisible" to interaction, and we shouldn't see events for enter/leave on the yellow box.

take this image for example:

![Image of cute frog](https://s-media-cache-ak0.pinimg.com/236x/86/d5/6c/86d56ce82a9310bad5c84932858ef325.jpg)

say that we have created this image with a single green background and various other basic shapes – circles, paths, rectangles, etc – gathered by a group. to allow interaction with the frog (shape collection), i may put  MouseEnter/Leave handlers _just on the green background_ – since that is the basic area of the image which defines all of the shapes. thus i can simply use the background to control interaction, perhaps to fade in/out the entire group.

what my simple example is demonstrating is that, if i mouse over *any other shape object* then that action will cycle enter/leave events, even if i'm not interested in those shapes (ie, no handlers on them).

to allow for this, any objects which don't have handlers should allow interaction to "fall through", as if they don't exist. 

in a way, i see this similar to event bubbling. if you 1. don't have a handler or 2. don't stop even propagation, then the event will continue up the parent chain. in this case, if an object isn't listening for mouse events then it shouldn't be an active target for them, allowing other objects under them to receive them.

**Comment by sasensi** [comment]
You can indeed workaround this issue by [locking](http://paperjs.org/reference/item/#locked) (this property is now documented) blue rectangle.
Here is a [sketch](http://sketch.paperjs.org/#V/0.11.8/S/pZJBa4NAEIX/yuBFBUnaQy+Gnkpu7aU5Nj1s1kkUtzOyOxpK8L93VxNNSimFiiu4vnnf+HZOEakPjPJoU6PoMsoizUV4Xy5BW1SCwB1ao5qmogNY1KLoYNBt7ZY6ZeETjeHj62UfHmFTqgYX005yClLw11ScA+ERZsVdBv6+fxgfaXYp2FfGPLFhm0M8cuLwqU9XF/rOtPg/doDO62d2oFyTw/L5KBGlS8AOScBUTpDQDsF8C2XB9MKtwzUJWt/lviUtFVMylKYwdamZHHu94UMSD+p4JParX2yfUXX4Z9tBfWN7/h3Duh4CncMCx1AJFIwOiAWq0FLbyPnUr4RlmBI/IcHr5lAWwRYL357YFgeeH7KdH626Ye/novztvf8C) demonstrating the solution.

```javascript
// create overlapping rectangles
var yellowRectangle = Shape.Rectangle({
    rectangle: new Rectangle(0, 0, 150, 150),
    fillColor: 'yellow'
});
var blueRectangle = Shape.Rectangle({
    rectangle: new Rectangle(50, 50, 50, 50),
    fillColor: 'blue'
});

// attach event listeners
yellowRectangle.onMouseEnter = function(event) {
    console.log('Enter');
};
yellowRectangle.onMouseLeave = function(event) {
    console.log('Leave');
};

// lock blue rectangle so it does not interupt yellow rectangle hovering
blueRectangle.locked = true;
```
